Docs 7754 dwformats duke (#1319)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/dataweave-formats-json.adoc

The DataWeave reader for JSON input supports the following parsing strategies:

* Indexed
* In-Memory
* Streaming

To understand the parsing strategies that DataWeave readers and writers can apply to this format, see xref:dataweave-formats.adoc#dw_readers_writers[DataWeave Parsing Strategies].

[[examples]]
== Examples

The following examples show uses of the JSON format:

* <<example1>>
* <<example2>>
* <<example3>>

[[example1]]
=== Example: Represent JSON in the DataWeave (dw) Format

This example shows how JSON input is represented in the DataWeave (dw) format.

==== Input

The following JSON object serves as the input payload to the DataWeave source.

[source,json,linenums]
----
{
  "name": "Matias",
  "age": 8,
  "male": true,
  "kids": null,
  "brothers": ["Pedro", "Sara"]
}
----

==== Source

The DataWeave script transforms the JSON input payload to the DataWeave (dw) format and MIME type.

[source,dataweave,linenums]
----
output application/dw
---
payload
----

==== Output

Notice that only the keys of the JSON input and the `application/dw` output differ. The JSON keys are surrounded by quotation marks. These DataWeave (dw)

> keys do not required quotation marks. See xref:dataweave-cookbook-extract-data.adoc#valid-keys[Valid Keys] for details.

Reviewer: ```suggestion
keys do not require quotation marks. See xref:dataweave-cookbook-extract-data.adoc#valid-keys[Valid Keys] for details.
```

[source,json,linenums]
----
[
  {
    "returnedElement": "95838"
  },
  {
    "returnedElement": "95823"
  },
  {
    "returnedElement": "95815"
  },
  {
    "returnedElement": "95815"
  },
  {
    "returnedElement": "95824"
  },
  {
    "returnedElement": "95841"
  }
]
----

// CONFIG PROPS ///////////////////////////////////////////////////////

[[properties]]
== Configuration Properties

DataWeave supports the following configuration properties for JSON.

=== Reader Properties (for JSON)

The JSON reader accepts properties that provide instructions for reading input data.

[cols="1,1,1,3a", options="header"]
|===
| Parameter | Type | Default | Description
| `streaming` | `Boolean` | `false` | Property for streaming input. Use only if
entries are accessed sequentially. Valid values are `true` or `false`.
The input must be a top-level array. For more on streaming in DataWeave, see
xref:dataweave-formats.adoc#dw_readers_writers[DataWeave Readers].
|===

=== Writer Properties (for JSON)

The JSON writer accepts properties that provide instructions for writing output data.

[cols="1,1,1,3a", options="header"]
|===
| Parameter | Type | Default | Description

> | `bufferSize` | `Number` | `8192` | Size of the buffer writer.

Reviewer: ```suggestion
| `bufferSize` | `Number` | `8192` | Size of the writer buffer.
```

> | `deferred` | `Boolean` | `false` | Property for deferred output.

Author: ```suggestion
| `deferred` | `Boolean` | `false` | Property for deferred output. Valid values are `true` or `false`.
```

| `duplicateKeyAsArray` | `Boolean` | `false` | If duplicate keys are detected

> in an object, the write will change the value to an array with all those values.

Reviewer: ```suggestion
in an object, the writer changes the value to an array with all those values.
```

Valid values are `true` or `false`. Note that JSON language does not allow
duplicate keys with one same parent, so the duplication usually raises an
exception.
| `encoding` | `String` | `UTF-8` | The character set to use for the output.
| `indent` | `Boolean` | `true` | Indicates whether to indent the JSON code for
better readability or to compress the JSON into a single line.
Valid values are `true` or `false`.

> | `skipNullOn` | `String` | None | Skips `null` values in the specified data

Author: ```suggestion
| `skipNullOn` | `String` | `null` | Skips `null` values in the specified data
```

> structure. By default it does not skip. Valid values are `elements`, `attributes`,

Author: ```suggestion
structure. By default, it does not skip the values. Valid values are `elements`, `attributes`,
```

or `everywhere`.

* `arrays` +
Ignore and omit `null` values from JSON output, for example, `output application/json skipNullOn="arrays"`.
* `objects`+
Ignore an object that has a null value. The output contains an empty object (`{}`) instead of the object with the null value, for example, `output application/json skipNullOn="objects"`.
* `everywhere` +
Apply `skipNullOn` to arrays and objects, for example, `output application/json skipNullOn="everywhere"`.

> |`writeAttributes` | `Boolean` | `false` | Indicates whether to add attributes of a key to children of the key. The new attribute key name will start with &amp;. Valid options are `true` or `false`.

Author: ```suggestion
|`writeAttributes` | `Boolean` | `false` | Indicates whether to add attributes of a key to children of the key. The new attribute key name will start with &#64;. Valid options are `true` or `false`.
```
